### PR TITLE
feat(eks): support adding k8s resources to imported clusters

### LINKED ISSUE #5383: [aws-eks] eks.Cluster.fromAttributes: Allow adding k8s resources on imported clusters
Currently I have an existing EKS cluster, is that possible to call `addResource` from another pipeline or someplace?

### Use Case

I tried to use [fromClusterAttributes](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-eks.Cluster.html#static-from-wbr-cluster-wbr-attributesscope-id-attrs-span-class-api-icon-api-icon-experimental-title-this-api-element-is-experimental-it-may-change-without-notice-span), but it only returns a [ImportedCluster](https://github.com/aws/aws-cdk/blob/88df1ebdbe9faf3717ed32a7daf96e3668bb483e/packages/%40aws-cdk/aws-eks/lib/cluster.ts#L748-L771) which does not have any static method.

It will be good to have:

```js
const cluster = new eks.Cluster.fromClusterAttributes(this, 'eks-cluster', {...});
cluster.addResource('ingress', {...});
cluster.addResource('service', {...});
```

### Other

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by eladb**
Allow adding Kubernetes resources such as manifests and Helm charts to imported clusters (`eks.Cluster.fromAttributes`).

To enable this behavior, when the cluster is imported, users will have to specify additional information:

 - `kubectlRole` - an IAM role that can issue kubectl commands against the cluster
 - `kubectlEnvironment` (optional) - environment variables for `kubectl`.
 - `kubectlPrivateSubnets` and `kubectlSecurityGroup` - required if the cluster's k8s endpoint is private

Resolves #5383

BREAKING CHANGE: when importing EKS clusters using `eks.Cluster.fromClusterAttributes`, the `clusterArn` attribute is not supported anymore, and will always be derived from `clusterName`.
* **eks**: Only a single `eks.Cluster` is allowed per CloudFormation stack.
* **eks**: The `securityGroups` attribute of `ClusterAttributes` is now `securityGroupIds`.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-eks/README.md [hunk 1]
```diff
@@ -428,6 +427,56 @@ Specifically, since the above use-case is quite common, there is an easier way t
...
+```
+
+At the minimum, when importing clusters for `kubectl` management, you will need
+to specify:
+
+- `clusterName` - the name of the cluster.
+- `kubectlRoleArn` - the ARN of an IAM role mapped to the `system:masters` RBAC
+  role. If the cluster you are importing was created using the AWS CDK, the
+  CloudFormation stack has an output that includes an IAM role that can be used.
+  Otherwise, you can create an IAM role and map it to `system:masters` manually.
```
[the remark this answers is not in the document]
Comment (eladb) [reply packages/@aws-cdk/aws-eks/README.md:470]: added

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 2]
```diff
@@ -77,44 +126,75 @@ export interface ICluster extends IResource, ec2.IConnectable {
...
   /**
    * The physical name of the Cluster
    */
   readonly clusterName: string;
 
-  /**
-   * The unique ARN assigned to the service by AWS
-   * in the form of arn:aws:eks:
-   */
-  readonly clusterArn: string;
```
[the remark this answers is not in the document]
Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:92]: yes, I guess.

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 3]
```diff
@@ -534,6 +653,41 @@ export class Cluster extends Resource implements ICluster {
    */
   public readonly defaultNodegroup?: Nodegroup;
 
+  /**
+   * An IAM role that can perform kubectl operations against this cluster.
+   *
+   * The role should be mapped to the `system:masters` Kubernetes RBAC role.
+   */
+  public readonly kubectlRole?: iam.IRole;
```
[the remark this answers is not in the document]
Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:661]: Correct, but the interface allows `undefined` so we can't change that (jsii forbids).

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 4]
```diff
@@ -534,6 +653,41 @@ export class Cluster extends Resource implements ICluster {
...
+   */
+  public readonly kubectlEnvironment?: { [key: string]: string };
+
+  /**
+   * A security group to use for `kubectl` execution.
+   *
+   * @default - If not specified, the k8s endpoint is expected to be accessible
+   * publicly.
+   */
+  public readonly kubectlSecurityGroup?: ec2.ISecurityGroup;
```
[the remark this answers is not in the document]
Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:675]: Same - the `ICluster` interface allows this to be undefined

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 5]
```diff
@@ -660,48 +812,32 @@ export class Cluster extends Resource implements ICluster {
...
+    this.kubectlSecurityGroup = new ec2.SecurityGroup(this, 'KubectlProviderSecurityGroup', {
       vpc: this.vpc,
       description: 'Comminication between KubectlProvider and EKS Control Plane',
     });
 
     // grant the kubectl provider access to the cluster control plane.
-    this.connections.allowFrom(this.kubctlProviderSecurityGroup, this.connections.defaultPort!);
+    this.connections.allowFrom(this.kubectlSecurityGroup, this.connections.defaultPort!);
+
+    this.adminRole = createAdminRole(this, {
```
[the remark this answers is not in the document]
Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:823]: Damn!

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 6]
```diff
@@ -1079,58 +1191,29 @@ export class Cluster extends Resource implements ICluster {
...
+  private defineKubectlProvider() {
+    const uid = '@aws-cdk/aws-eks.KubectlProvider';
 
-    if (!this._kubectlReadyBarrier) {
-      throw new Error('unexpected: kubectl enabled clusters should have a kubectl-ready barrier resource');
+    // since we can't have the provider connect to multiple networks, and we
+    // wanted to avoid resource tear down, we decided for now that we will only
+    // support a single EKS cluster per CFN stack.
+    if (this.stack.node.tryFindChild(uid)) {
+      throw new Error('Only a single EKS cluster can be defined within a CloudFormation stack');
```
[the remark this answers is not in the document]
Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:1210]: Done

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 7]
```diff
@@ -77,44 +126,75 @@ export interface ICluster extends IResource, ec2.IConnectable {
...
+   */
+  readonly securityGroupIds?: string[];
 
   /**
-   * The security groups associated with this cluster.
+   * An IAM role with cluster administrator and "system:masters" permissions.
+   * @default - if not specified, it not be possible to issue `kubectl` commands
+   * against an imported cluster.
    */
-  readonly securityGroups: ec2.ISecurityGroup[];
```
[the remark this answers is not in the document]
Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:117]: added

File: packages/@aws-cdk/aws-eks/lib/kubectl-provider.ts [hunk 8]
```diff
@@ -1,55 +1,70 @@
...
-   */
-  readonly securityGroups?: ISecurityGroup[];
+    // if this is an "owned" cluster, it has a provider associated with it
+    if (cluster instanceof Cluster) {
+      return cluster._attachKubectlResourceScope(scope);
+    }
+
+    if (!cluster.kubectlRole) {
+      throw new Error('cluster does not have "kubectlRole" defined, so "kubectl" commands cannot be issued against it');
+    }
```
[the remark this answers is not in the document]
Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/kubectl-provider.ts:26]: removed from here

File: packages/@aws-cdk/aws-eks/lib/cluster.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts]: Okay, fixed.

File: packages/@aws-cdk/aws-eks/lib/cluster.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (iliapolo) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:660]: Ah, good to know 👍

File: packages/@aws-cdk/aws-eks/lib/cluster.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (iliapolo) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:674]: 👍

**Review by iliapolo — APPROVED** [review]

**Comment by iliapolo** [comment]
@Mergifyio refresh

**Comment by iliapolo** [comment]
@eladb Any reason the `do-not-merge` label is still here?

**Comment by eladb** [comment]
@mergifyio refresh

**Comment by pahud** [comment]
This is amazing! Thank you for the great work!

**Comment by eladb** [comment]
> This is amazing! Thank you for the great work!

Thanks! Looking forward for feedback :-)

**Comment by itajaja** [comment]
hey, just updated. we have two clusters in our stack. does this mean we cannot update, or is there a way to go around this?

**Comment by iliapolo** [comment]
@itajaja This limitation actually short-circuits some problems you might encounter with multiple clusters in a stack with regards to endpoint access configuration. Is it possible for you to split the cluster to different stacks?

**Comment by itajaja** [comment]
what kinds of problems? are they solvable through aws-cdk alone?

> Is it possible for you to split the cluster to different stacks?

it would be quite awkward.

**Comment by iliapolo** [comment]
@itajaja  

> what kinds of problems? are they solvable through aws-cdk alone?

Currently no. The problem is that having multiple clusters in the same stack would force those clusters to have a publicly accessible endpoint (i.e only `EndpointAccess.PUBLIC`).

Apologies for the inconvenience, we have created an [issue](https://github.com/aws/aws-cdk/issues/10073) to gather feedback on this limitation and think of possible solutions. Would be great if you can share your use case.

**Pull request merged**